### ISSUE TITLE
Length of String is wrong.

### ISSUE THREAD

**Issue by watiko** [nodegit/nodegit#785]
NodeGit uses substring with bytes length comes from strlen but JavaScript requires characters length.
So non-ascii string will be broken.

https://libgit2.github.com/libgit2/#HEAD/type/git_diff_line
git_diff_line->content_len : Number of bytes of data

NodeGit:
https://github.com/nodegit/nodegit/blob/v0.5.0/lib/convenient_line.js#L71

Some other implementations:
https://github.com/libgit2/rugged/blob/v0.23.3/ext/rugged/rugged_diff_line.c#L68
https://github.com/libgit2/libgit2sharp/blob/v0.21.1/LibGit2Sharp/Core/EncodingMarshaler.cs#L129
https://github.com/libgit2/objective-git/blob/0.8.8/ObjectiveGit/GTDiffLine.m#L22

Thx.

**Comment by maxkorp** [comment]
Good catch!

**Comment by watiko** [comment]
I wrote a following ugly fix in the layer of JavaScript.
But I think this issue should be resolved in that of C++.

```diff
--- a/convenient_line.js	2015-11-13 23:47:56.299338415 +0900
+++ b/convenient_line.js	2015-11-13 23:48:19.766911535 +0900
@@ -67,8 +67,22 @@
 * @return {String}
 */
 ConvenientLine.prototype.content = function() {
-  return this.raw.content()
-    .substring(0, this.raw.contentLen()).replace("\n", "");
+  function countBytes(chr) {
+      return (encodeURIComponent(chr).replace(/%../g,"x").length);
+  }
+
+  var rawContent = this.raw.content();
+  var rawContentLen = this.raw.contentLen();
+  var trueContentLen = 0;
+
+  rawContent.split('').some(function(chr) {
+    rawContentLen -= countBytes(chr);
+    trueContentLen += 1;
+    if (rawContentLen == 0)
+      return true;
+  });
+
+  return rawContent.substring(0, trueContentLen).replace("\n", "");
 };
 
 NodeGit.ConvenientLine = ConvenientLine;
```

**Comment by johnhaley81** [comment]
@watiko would you mind PR'ing this? The problem with implementing this in C++ is that we'd have to figure out way to generate that code in our generators. I'd argue that you *should* be using the JS convenience wrappers/functions unless you really know what you're doing and want that freedom.

Also, could you cache this value? We should need to recalculate this stuff since it's immutable.

**Comment by watiko** [comment]
I did not mean it for PR'ing. It is only ploof of concept that this issue can resolve in the layer of JavaScript. And I don't know the structure of NodeGit very much so I couldn't write patch correcty sorry.

**Comment by johnhaley81** [comment]
Closed via #844
